sympy/core/function.py

```diff
@@ -823,13 +835,13 @@ class AppliedUndef(Function):
 
     is_number = False
 
-    def __new__(cls, *args, **options):
-        args = list(map(sympify, args))
+    def __new__(cls, *args, **options) -> Expr:  # type: ignore
+        args = tuple(map(sympify, args))
         u = [a.name for a in args if isinstance(a, UndefinedFunction)]
         if u:
             raise TypeError('Invalid argument: expecting an expression, not UndefinedFunction%s: %s' % (
                 's'*(len(u) > 1), ', '.join(u)))
-        obj = super().__new__(cls, *args, **options)
+        obj: Expr = super().__new__(cls, *args, **options)  # type: ignore
```

Comment:
Maybe `_new_` should be used directly from here.

Reply:
The `type: ignore` here is because mypy rejects having a `cls.__new__` method whose return hint does not imply that it returns an instance of `cls`:
```python
from __future__ import annotations

class A:
    def __new__(cls) -> A:
        return super().__new__(cls)

class B(A):
    def __new__(cls) -> A:
        return super().__new__(cls)

reveal_type(B())
```
Here mypy rejects this although pyright accepts it:
```console
$ mypy q.py 
q.py:8: error: Incompatible return type for "__new__" (returns "A", but must return a subtype of "B")  [misc]
q.py:11: note: Revealed type is "q.B"
Found 1 error in 1 file (checked 1 source file)
$ pyright q.py 
  q.py:11:13 - information: Type of "B()" is "A"
0 errors, 0 warnings, 1 information 
```
I could change this to say that the return type is `AppliedUndef` and then mypy would not complain. I am not sure if this should really be considered a bug in mypy though.

The return type `Expr` was needed for `Function.__new__` because you can have e.g. `sin(pi) -> 0` so we have no guarantee that `sin(obj)` is of type `sin` but it should always be of type `Expr`. With `AppliedUndef` there is no evaluation though so it would be accurate to say that it returns `AppliedUndef`. I'm just not sure if that is more useful to users than `Expr` (which is not incorrect since `AppliedUndef` is a subclass of `Expr`).